relational_operators/WorkOrder.proto

```diff
@@ -0,0 +1,244 @@
+//   Copyright 2011-2015 Quickstep Technologies LLC.
+//   Copyright 2015-2016 Pivotal Software, Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+syntax = "proto2";
+
+package quickstep.serialization;
+
+import "relational_operators/SortMergeRunOperator.proto";
+import "relational_operators/TextScanOperator.proto";
+
+enum WorkOrderType {
+  AGGREGATION = 1;
+  BUILD_HASH = 2;
+  CREATE_INDEX = 3;  // Placeholder.
+  CREATE_TABLE = 4;  // Placeholder.
+  DELETE = 5;
+  DESTROY_HASH = 6;
+  DROP_TABLE = 7;
+  FINALIZE_AGGREGATION = 8;
+  HASH_JOIN = 9;
+  INSERT = 10;
+  NESTED_LOOP_JOIN = 11;
+  SAMPLE = 12;
+  SAVE_BLOCKS = 13;
+  SELECT = 14;
+  SORT_MERGE_RUN = 15;
+  SORT_RUN_GENERATION = 16;
+  TABLE_GENERATOR = 17;
+  TEXT_SCAN = 18;
+  TEXT_SPLIT = 19;
+  UPDATE = 20;
+}
+
+message WorkOrder {
+  required WorkOrderType work_order_type = 1;
+
+  // The convention for extension numbering is that extensions for a particular
+  // WorkOrderID should begin from (operator_type + 1) * 16.
+  extensions 16 to max;
+}
+
+message AggregationWorkOrder {
+  extend WorkOrder {
+    // All required.
+    optional uint32 aggr_state_index = 16;
+    optional fixed64 block_id = 17;
+  }
+}
+
+message BuildHashWorkOrder {
+  extend WorkOrder {
+    // All required.
+    optional int32 relation_id = 32;
+    repeated int32 join_key_attributes = 33;
+    optional bool any_join_key_attributes_nullable = 34;
+    optional uint32 join_hash_table_index = 35;
+    optional fixed64 block_id = 36;
+  }
+}
+
+message DeleteWorkOrder {
+  extend WorkOrder {
+    // All required.
+    optional uint64 operator_index = 96;
+    optional int32 relation_id = 97;
+    optional int32 predicate_index = 98;
+    optional fixed64 block_id = 99;
+  }
+}
+
+message DestroyHashWorkOrder {
+  extend WorkOrder {
+    // All required.
+    optional uint32 join_hash_table_index = 112;
+  }
+}
+
+message DropTableWorkOrder {
+  extend WorkOrder {
+    // If set, CatalogDatabaseCache will drop the relation.
+    optional int32 relation_id = 128;
+    // Optional, and maybe empty.
+    repeated fixed64 block_ids = 129;
+  }
+}
+
+message FinalizeAggregationWorkOrder {
+  extend WorkOrder {
+    // All required.
+    optional uint32 aggr_state_index = 144;
+    optional int32 insert_destination_index = 145;
```

Comment:
So if someone creates a (fictional) ``AaaWorkOrder``, they will have to renumber 16 to 341 for no good reason. I would strongly oppose the sorting requirement.